### ISSUE TITLE
ApplicationLogs and Neo3 GAS transfer notification results

### ISSUE THREAD

**Issue by hal0x2328** [neo-project/neo#1252]
The application log for a GAS transfer shows only the from/to/amount transfer notification for GAS transfers - but GAS is a special case, since the system and network fee is subtracted from that amount, so any trackers using transfer notifications to tally balances or confirm transfer amounts for a dApp will end up with incorrect results.

There is the `gas_consumed` property of the application log, but it doesn't reflect the system fee, network fee (or the sum of both) either so it can't be used to arrive at the correct result. 

The only way currently to correctly calculate the amount of the GAS transfer from address A to address B is to also request the raw transaction and subtract the fees shown there.

Example on Neo3 TestNet, a transfer of 1.9770058 GAS:
```
neo> send gas AHpP2cbWfUDBuH2UVKbmLsc6AHEYYG92Vn 1.9770058
password: *****************
TXID: 0xbbc320ceb17c704987deb0db157a10248f983b06969467e51a6b960c9bc02efc
```
The resulting log:
```
{
    "jsonrpc": "2.0",
    "id": 1,
    "result": {
        "txid": "0xbbc320ceb17c704987deb0db157a10248f983b06969467e51a6b960c9bc02efc",
        "trigger": "Application",
        "vmstate": "HALT",
        "gas_consumed": "9007660",
        "stack": [],
        "notifications": [
            {
                "contract": "0xa1760976db5fcdfab2a9930e8f6ce875b2d18225",
                "state": {
                    "type": "Array",
                    "value": [
                        {
                            "type": "ByteArray",
                            "value": "5472616e73666572"
                        },
                        {
                            "type": "ByteArray",
                            "value": "78e2f3100f2c76823345835c96b8f3b1897a7256"
                        },
                        {
                            "type": "ByteArray",
                            "value": "16699c1a5374ec7b5ce880449e14e7c99ef3fd24"
                        },
                        {
                            "type": "Integer",
                            "value": "197700580"
                        }
                    ]
                }
            }
        ]
    }
}
```
From the application log, it *looks* like 16699c1a5374ec7b5ce880449e14e7c99ef3fd24 received 1.97700580 GAS but there was actually a combined fee of 1.0340366 GAS subtracted so actually in the end 16699c1a5374ec7b5ce880449e14e7c99ef3fd24's balance only increased by 0.9429692 GAS.

**Comment by shargon** [comment]
Maybe this will be fixed with https://github.com/neo-project/neo/pull/1248

**Comment by shargon** [comment]
@tommo-L could you check if your patch https://github.com/neo-project/neo/pull/1248 fizbthis issue too?

**Comment by Tommo-L** [comment]
I think this is another problem, I'll check it.

**Comment by Tommo-L** [comment]
@hal0x2328 
Is it wrong?

![image](https://user-images.githubusercontent.com/43396869/69798796-f191e780-120c-11ea-959a-9df49204384f.png)

>neo> send gas AHpP2cbWfUDBuH2UVKbmLsc6AHEYYG92Vn 1.9770058
password: *****************
TXID: 0xbbc320ceb17c704987deb0db157a10248f983b06969467e51a6b960c9bc02efc

**Comment by Tommo-L** [comment]
I got the point, there is no bug. I think the two addresses: `AHpP2cbWfUDBuH2UVKbmLsc6AHEYYG92Vn` and `ASo4cF4GPWSMAT7JYvoYe9G5BWMyrsKtcn` are in the one wallet.

```
neo> send gas AHpP2cbWfUDBuH2UVKbmLsc6AHEYYG92Vn 1.9770058
password: *****************
TXID: 0xbbc320ceb17c704987deb0db157a10248f983b06969467e51a6b960c9bc02efc
```

https://neo3-preview.com/transaction/0xbbc320ceb17c704987deb0db157a10248f983b06969467e51a6b960c9bc02efc

As you can see, the tx's sender is `AHpP2cbWfUDBuH2UVKbmLsc6AHEYYG92Vn`, so the tx's fee = `1.0340366` was paied from `AHpP2cbWfUDBuH2UVKbmLsc6AHEYYG92Vn`.
Then the `1.97700580` was sent from `ASo4cF4GPWSMAT7JYvoYe9G5BWMyrsKtcn` actually.
At least, `AHpP2cbWfUDBuH2UVKbmLsc6AHEYYG92Vn` cost `1.0340366` and recevied `1.97700580`, the income is:
```
-1.0340366 +  1.97700580 =  0.9429692
```

**Comment by hal0x2328** [comment]
This transaction only sent part of `AHpP2cbWfUDBuH2UVKbmLsc6AHEYYG92Vn`'s balance. The bug is exhibited only when you send the whole balance of an address. 

So if AHpP2cbWfUDBuH2UVKbmLsc6AHEYYG92Vn now sends the entire current balance of 49995.919995 GAS to another address, the fee must be subtracted from the 49995.919995 - the receiving address will get the remainder which will be some amount less than 49995.919995. This behavior is expected and correct even if it makes it a bit unpredictable to the sender as to the final amount the recipient will receive. 

The issue I saw is that the `Transfer` notification event will show that the full 49995.919995 was sent to the recipient address instead of the correct amount. So an explorer that is tracking the transfer via the notification will have the wrong information.

**Comment by Tommo-L** [comment]
> So if AHpP2cbWfUDBuH2UVKbmLsc6AHEYYG92Vn now sends the entire current balance of 49995.919995 GAS to another address, the fee must be subtracted from the 49995.919995 - the receiving address will get the remainder which will be some amount less than 49995.919995.

I will have a try, but I think the tx will fail as `Insufficient GAS`. Even though it can be packaged in the blockchain, but it will execute fail.

**Comment by hal0x2328** [comment]
Would that mean than once an account has some amount of GAS, it can never have a zero balance again? i.e. there will always be some fractional "dust" GAS that can't be sent out because it's less than the minimum fee to send it.

**Comment by hal0x2328** [comment]
I see it will be possible with Neo3 to have "dust" GAS in an account that can't be transferred out because the transaction fee is more than the amount it would cost to send it. They could transfer in more GAS and send the properly calculated amount + fee to leave the amount at exactly zero but it's going to lead to confusion either way - wallets will need to be smart about handling it.

**Comment by Tommo-L** [comment]
> it will be possible with Neo3 to have "dust" GAS in an account that can't be transferred out because the transaction fee is more than the amount it would cost to send it. 

This was fixed in #1248. It works fine my latest version
![image](https://user-images.githubusercontent.com/43396869/69949538-ac332a00-152c-11ea-96e3-0da955dfe69b.png)

What surprised me was that the transaction with insufficient balance will execute successfully
